// Code generated by fhirgen v0.2.0. DO NOT EDIT.
// Generated at: 2026-02-15T12:54:49Z
// FHIR Version: R4
// Source: FHIR StructureDefinitions from https://hl7.org/fhir/R4/

package r4

import (
	"github.com/zs-health/zh-fhir-go/fhir"
	"github.com/zs-health/zh-fhir-go/fhir/primitives"
)

// ResourceTypeMolecularSequence is the FHIR resource type name for MolecularSequence.
const ResourceTypeMolecularSequence = "MolecularSequence"

// MolecularSequenceReferenceSeq represents a FHIR BackboneElement for MolecularSequence.referenceSeq.
type MolecularSequenceReferenceSeq struct {
	// Unique id for inter-element referencing
	ID *string `json:"id,omitempty" fhir:"cardinality=0..1"`
	// Extension for ID
	IDExt *primitives.PrimitiveExtension `json:"_id,omitempty" fhir:"cardinality=0..1"`
	// Additional content defined by implementations
	Extension []Extension `json:"extension,omitempty" fhir:"cardinality=0..*"`
	// Extensions that cannot be ignored even if unrecognized
	ModifierExtension []Extension `json:"modifierExtension,omitempty" fhir:"cardinality=0..*,summary"`
	// Chromosome containing genetic finding
	Chromosome *CodeableConcept `json:"chromosome,omitempty" fhir:"cardinality=0..1,summary"`
	// The Genome Build used for reference, following GRCh build versions e.g. 'GRCh 37'
	GenomeBuild *string `json:"genomeBuild,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for GenomeBuild
	GenomeBuildExt *primitives.PrimitiveExtension `json:"_genomeBuild,omitempty" fhir:"cardinality=0..1"`
	// sense | antisense
	Orientation *string `json:"orientation,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Orientation
	OrientationExt *primitives.PrimitiveExtension `json:"_orientation,omitempty" fhir:"cardinality=0..1"`
	// Reference identifier
	ReferenceSeqId *CodeableConcept `json:"referenceSeqId,omitempty" fhir:"cardinality=0..1,summary"`
	// A pointer to another MolecularSequence entity as reference sequence
	ReferenceSeqPointer *Reference `json:"referenceSeqPointer,omitempty" fhir:"cardinality=0..1,summary"`
	// A string to represent reference sequence
	ReferenceSeqString *string `json:"referenceSeqString,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for ReferenceSeqString
	ReferenceSeqStringExt *primitives.PrimitiveExtension `json:"_referenceSeqString,omitempty" fhir:"cardinality=0..1"`
	// watson | crick
	Strand *string `json:"strand,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Strand
	StrandExt *primitives.PrimitiveExtension `json:"_strand,omitempty" fhir:"cardinality=0..1"`
	// Start position of the window on the  reference sequence
	WindowStart *int `json:"windowStart,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for WindowStart
	WindowStartExt *primitives.PrimitiveExtension `json:"_windowStart,omitempty" fhir:"cardinality=0..1"`
	// End position of the window on the reference sequence
	WindowEnd *int `json:"windowEnd,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for WindowEnd
	WindowEndExt *primitives.PrimitiveExtension `json:"_windowEnd,omitempty" fhir:"cardinality=0..1"`
}

// MolecularSequenceVariant represents a FHIR BackboneElement for MolecularSequence.variant.
type MolecularSequenceVariant struct {
	// Unique id for inter-element referencing
	ID *string `json:"id,omitempty" fhir:"cardinality=0..1"`
	// Extension for ID
	IDExt *primitives.PrimitiveExtension `json:"_id,omitempty" fhir:"cardinality=0..1"`
	// Additional content defined by implementations
	Extension []Extension `json:"extension,omitempty" fhir:"cardinality=0..*"`
	// Extensions that cannot be ignored even if unrecognized
	ModifierExtension []Extension `json:"modifierExtension,omitempty" fhir:"cardinality=0..*,summary"`
	// Start position of the variant on the  reference sequence
	Start *int `json:"start,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Start
	StartExt *primitives.PrimitiveExtension `json:"_start,omitempty" fhir:"cardinality=0..1"`
	// End position of the variant on the reference sequence
	End *int `json:"end,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for End
	EndExt *primitives.PrimitiveExtension `json:"_end,omitempty" fhir:"cardinality=0..1"`
	// Allele that was observed
	ObservedAllele *string `json:"observedAllele,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for ObservedAllele
	ObservedAlleleExt *primitives.PrimitiveExtension `json:"_observedAllele,omitempty" fhir:"cardinality=0..1"`
	// Allele in the reference sequence
	ReferenceAllele *string `json:"referenceAllele,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for ReferenceAllele
	ReferenceAlleleExt *primitives.PrimitiveExtension `json:"_referenceAllele,omitempty" fhir:"cardinality=0..1"`
	// Extended CIGAR string for aligning the sequence with reference bases
	Cigar *string `json:"cigar,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Cigar
	CigarExt *primitives.PrimitiveExtension `json:"_cigar,omitempty" fhir:"cardinality=0..1"`
	// Pointer to observed variant information
	VariantPointer *Reference `json:"variantPointer,omitempty" fhir:"cardinality=0..1,summary"`
}

// MolecularSequenceQualityRoc represents a FHIR BackboneElement for MolecularSequence.quality.roc.
type MolecularSequenceQualityRoc struct {
	// Unique id for inter-element referencing
	ID *string `json:"id,omitempty" fhir:"cardinality=0..1"`
	// Extension for ID
	IDExt *primitives.PrimitiveExtension `json:"_id,omitempty" fhir:"cardinality=0..1"`
	// Additional content defined by implementations
	Extension []Extension `json:"extension,omitempty" fhir:"cardinality=0..*"`
	// Extensions that cannot be ignored even if unrecognized
	ModifierExtension []Extension `json:"modifierExtension,omitempty" fhir:"cardinality=0..*,summary"`
	// Genotype quality score
	Score []int `json:"score,omitempty" fhir:"cardinality=0..*,summary"`
	// Extension for Score
	ScoreExt *primitives.PrimitiveExtension `json:"_score,omitempty" fhir:"cardinality=0..1"`
	// Roc score true positive numbers
	NumTP []int `json:"numTP,omitempty" fhir:"cardinality=0..*,summary"`
	// Extension for NumTP
	NumTPExt *primitives.PrimitiveExtension `json:"_numTP,omitempty" fhir:"cardinality=0..1"`
	// Roc score false positive numbers
	NumFP []int `json:"numFP,omitempty" fhir:"cardinality=0..*,summary"`
	// Extension for NumFP
	NumFPExt *primitives.PrimitiveExtension `json:"_numFP,omitempty" fhir:"cardinality=0..1"`
	// Roc score false negative numbers
	NumFN []int `json:"numFN,omitempty" fhir:"cardinality=0..*,summary"`
	// Extension for NumFN
	NumFNExt *primitives.PrimitiveExtension `json:"_numFN,omitempty" fhir:"cardinality=0..1"`
	// Precision of the GQ score
	Precision []float64 `json:"precision,omitempty" fhir:"cardinality=0..*,summary"`
	// Extension for Precision
	PrecisionExt *primitives.PrimitiveExtension `json:"_precision,omitempty" fhir:"cardinality=0..1"`
	// Sensitivity of the GQ score
	Sensitivity []float64 `json:"sensitivity,omitempty" fhir:"cardinality=0..*,summary"`
	// Extension for Sensitivity
	SensitivityExt *primitives.PrimitiveExtension `json:"_sensitivity,omitempty" fhir:"cardinality=0..1"`
	// FScore of the GQ score
	FMeasure []float64 `json:"fMeasure,omitempty" fhir:"cardinality=0..*,summary"`
	// Extension for FMeasure
	FMeasureExt *primitives.PrimitiveExtension `json:"_fMeasure,omitempty" fhir:"cardinality=0..1"`
}

// MolecularSequenceQuality represents a FHIR BackboneElement for MolecularSequence.quality.
type MolecularSequenceQuality struct {
	// Unique id for inter-element referencing
	ID *string `json:"id,omitempty" fhir:"cardinality=0..1"`
	// Extension for ID
	IDExt *primitives.PrimitiveExtension `json:"_id,omitempty" fhir:"cardinality=0..1"`
	// Additional content defined by implementations
	Extension []Extension `json:"extension,omitempty" fhir:"cardinality=0..*"`
	// Extensions that cannot be ignored even if unrecognized
	ModifierExtension []Extension `json:"modifierExtension,omitempty" fhir:"cardinality=0..*,summary"`
	// indel | snp | unknown
	Type string `json:"type" fhir:"cardinality=1..1,required,summary"`
	// Extension for Type
	TypeExt *primitives.PrimitiveExtension `json:"_type,omitempty" fhir:"cardinality=0..1"`
	// Standard sequence for comparison
	StandardSequence *CodeableConcept `json:"standardSequence,omitempty" fhir:"cardinality=0..1,summary"`
	// Start position of the sequence
	Start *int `json:"start,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Start
	StartExt *primitives.PrimitiveExtension `json:"_start,omitempty" fhir:"cardinality=0..1"`
	// End position of the sequence
	End *int `json:"end,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for End
	EndExt *primitives.PrimitiveExtension `json:"_end,omitempty" fhir:"cardinality=0..1"`
	// Quality score for the comparison
	Score *Quantity `json:"score,omitempty" fhir:"cardinality=0..1,summary"`
	// Method to get quality
	Method *CodeableConcept `json:"method,omitempty" fhir:"cardinality=0..1,summary"`
	// True positives from the perspective of the truth data
	TruthTP *float64 `json:"truthTP,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for TruthTP
	TruthTPExt *primitives.PrimitiveExtension `json:"_truthTP,omitempty" fhir:"cardinality=0..1"`
	// True positives from the perspective of the query data
	QueryTP *float64 `json:"queryTP,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for QueryTP
	QueryTPExt *primitives.PrimitiveExtension `json:"_queryTP,omitempty" fhir:"cardinality=0..1"`
	// False negatives
	TruthFN *float64 `json:"truthFN,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for TruthFN
	TruthFNExt *primitives.PrimitiveExtension `json:"_truthFN,omitempty" fhir:"cardinality=0..1"`
	// False positives
	QueryFP *float64 `json:"queryFP,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for QueryFP
	QueryFPExt *primitives.PrimitiveExtension `json:"_queryFP,omitempty" fhir:"cardinality=0..1"`
	// False positives where the non-REF alleles in the Truth and Query Call Sets match
	GtFP *float64 `json:"gtFP,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for GtFP
	GtFPExt *primitives.PrimitiveExtension `json:"_gtFP,omitempty" fhir:"cardinality=0..1"`
	// Precision of comparison
	Precision *float64 `json:"precision,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Precision
	PrecisionExt *primitives.PrimitiveExtension `json:"_precision,omitempty" fhir:"cardinality=0..1"`
	// Recall of comparison
	Recall *float64 `json:"recall,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Recall
	RecallExt *primitives.PrimitiveExtension `json:"_recall,omitempty" fhir:"cardinality=0..1"`
	// F-score
	FScore *float64 `json:"fScore,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for FScore
	FScoreExt *primitives.PrimitiveExtension `json:"_fScore,omitempty" fhir:"cardinality=0..1"`
	// Receiver Operator Characteristic (ROC) Curve
	Roc *MolecularSequenceQualityRoc `json:"roc,omitempty" fhir:"cardinality=0..1,summary"`
}

// MolecularSequenceRepository represents a FHIR BackboneElement for MolecularSequence.repository.
type MolecularSequenceRepository struct {
	// Unique id for inter-element referencing
	ID *string `json:"id,omitempty" fhir:"cardinality=0..1"`
	// Extension for ID
	IDExt *primitives.PrimitiveExtension `json:"_id,omitempty" fhir:"cardinality=0..1"`
	// Additional content defined by implementations
	Extension []Extension `json:"extension,omitempty" fhir:"cardinality=0..*"`
	// Extensions that cannot be ignored even if unrecognized
	ModifierExtension []Extension `json:"modifierExtension,omitempty" fhir:"cardinality=0..*,summary"`
	// directlink | openapi | login | oauth | other
	Type string `json:"type" fhir:"cardinality=1..1,required,summary"`
	// Extension for Type
	TypeExt *primitives.PrimitiveExtension `json:"_type,omitempty" fhir:"cardinality=0..1"`
	// URI of the repository
	URL *string `json:"url,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for URL
	URLExt *primitives.PrimitiveExtension `json:"_url,omitempty" fhir:"cardinality=0..1"`
	// Repository's name
	Name *string `json:"name,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Name
	NameExt *primitives.PrimitiveExtension `json:"_name,omitempty" fhir:"cardinality=0..1"`
	// Id of the dataset that used to call for dataset in repository
	DatasetId *string `json:"datasetId,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for DatasetId
	DatasetIdExt *primitives.PrimitiveExtension `json:"_datasetId,omitempty" fhir:"cardinality=0..1"`
	// Id of the variantset that used to call for variantset in repository
	VariantsetId *string `json:"variantsetId,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for VariantsetId
	VariantsetIdExt *primitives.PrimitiveExtension `json:"_variantsetId,omitempty" fhir:"cardinality=0..1"`
	// Id of the read
	ReadsetId *string `json:"readsetId,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for ReadsetId
	ReadsetIdExt *primitives.PrimitiveExtension `json:"_readsetId,omitempty" fhir:"cardinality=0..1"`
}

// MolecularSequenceStructureVariantOuter represents a FHIR BackboneElement for MolecularSequence.structureVariant.outer.
type MolecularSequenceStructureVariantOuter struct {
	// Unique id for inter-element referencing
	ID *string `json:"id,omitempty" fhir:"cardinality=0..1"`
	// Extension for ID
	IDExt *primitives.PrimitiveExtension `json:"_id,omitempty" fhir:"cardinality=0..1"`
	// Additional content defined by implementations
	Extension []Extension `json:"extension,omitempty" fhir:"cardinality=0..*"`
	// Extensions that cannot be ignored even if unrecognized
	ModifierExtension []Extension `json:"modifierExtension,omitempty" fhir:"cardinality=0..*,summary"`
	// Structural variant outer start
	Start *int `json:"start,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Start
	StartExt *primitives.PrimitiveExtension `json:"_start,omitempty" fhir:"cardinality=0..1"`
	// Structural variant outer end
	End *int `json:"end,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for End
	EndExt *primitives.PrimitiveExtension `json:"_end,omitempty" fhir:"cardinality=0..1"`
}

// MolecularSequenceStructureVariantInner represents a FHIR BackboneElement for MolecularSequence.structureVariant.inner.
type MolecularSequenceStructureVariantInner struct {
	// Unique id for inter-element referencing
	ID *string `json:"id,omitempty" fhir:"cardinality=0..1"`
	// Extension for ID
	IDExt *primitives.PrimitiveExtension `json:"_id,omitempty" fhir:"cardinality=0..1"`
	// Additional content defined by implementations
	Extension []Extension `json:"extension,omitempty" fhir:"cardinality=0..*"`
	// Extensions that cannot be ignored even if unrecognized
	ModifierExtension []Extension `json:"modifierExtension,omitempty" fhir:"cardinality=0..*,summary"`
	// Structural variant inner start
	Start *int `json:"start,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Start
	StartExt *primitives.PrimitiveExtension `json:"_start,omitempty" fhir:"cardinality=0..1"`
	// Structural variant inner end
	End *int `json:"end,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for End
	EndExt *primitives.PrimitiveExtension `json:"_end,omitempty" fhir:"cardinality=0..1"`
}

// MolecularSequenceStructureVariant represents a FHIR BackboneElement for MolecularSequence.structureVariant.
type MolecularSequenceStructureVariant struct {
	// Unique id for inter-element referencing
	ID *string `json:"id,omitempty" fhir:"cardinality=0..1"`
	// Extension for ID
	IDExt *primitives.PrimitiveExtension `json:"_id,omitempty" fhir:"cardinality=0..1"`
	// Additional content defined by implementations
	Extension []Extension `json:"extension,omitempty" fhir:"cardinality=0..*"`
	// Extensions that cannot be ignored even if unrecognized
	ModifierExtension []Extension `json:"modifierExtension,omitempty" fhir:"cardinality=0..*,summary"`
	// Structural variant change type
	VariantType *CodeableConcept `json:"variantType,omitempty" fhir:"cardinality=0..1,summary"`
	// Does the structural variant have base pair resolution breakpoints?
	Exact *bool `json:"exact,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Exact
	ExactExt *primitives.PrimitiveExtension `json:"_exact,omitempty" fhir:"cardinality=0..1"`
	// Structural variant length
	Length *int `json:"length,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Length
	LengthExt *primitives.PrimitiveExtension `json:"_length,omitempty" fhir:"cardinality=0..1"`
	// Structural variant outer
	Outer *MolecularSequenceStructureVariantOuter `json:"outer,omitempty" fhir:"cardinality=0..1,summary"`
	// Structural variant inner
	Inner *MolecularSequenceStructureVariantInner `json:"inner,omitempty" fhir:"cardinality=0..1,summary"`
}

// MolecularSequence represents a FHIR MolecularSequence.
type MolecularSequence struct {
	fhir.DomainResource
	// Extension for Contained
	ContainedExt *primitives.PrimitiveExtension `json:"_contained,omitempty" fhir:"cardinality=0..1"`
	// Unique ID for this particular sequence. This is a FHIR-defined id
	Identifier []Identifier `json:"identifier,omitempty" fhir:"cardinality=0..*,summary"`
	// aa | dna | rna
	Type *string `json:"type,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for Type
	TypeExt *primitives.PrimitiveExtension `json:"_type,omitempty" fhir:"cardinality=0..1"`
	// Base number of coordinate system (0 for 0-based numbering or coordinates, inclusive start, exclusive end, 1 for 1-based numbering, inclusive start, inclusive end)
	CoordinateSystem int `json:"coordinateSystem" fhir:"cardinality=1..1,required,summary"`
	// Extension for CoordinateSystem
	CoordinateSystemExt *primitives.PrimitiveExtension `json:"_coordinateSystem,omitempty" fhir:"cardinality=0..1"`
	// Who and/or what this is about
	Patient *Reference `json:"patient,omitempty" fhir:"cardinality=0..1,summary"`
	// Specimen used for sequencing
	Specimen *Reference `json:"specimen,omitempty" fhir:"cardinality=0..1,summary"`
	// The method for sequencing
	Device *Reference `json:"device,omitempty" fhir:"cardinality=0..1,summary"`
	// Who should be responsible for test result
	Performer *Reference `json:"performer,omitempty" fhir:"cardinality=0..1,summary"`
	// The number of copies of the sequence of interest.  (RNASeq)
	Quantity *Quantity `json:"quantity,omitempty" fhir:"cardinality=0..1,summary"`
	// A sequence used as reference
	ReferenceSeq *MolecularSequenceReferenceSeq `json:"referenceSeq,omitempty" fhir:"cardinality=0..1,summary"`
	// Variant in sequence
	Variant []MolecularSequenceVariant `json:"variant,omitempty" fhir:"cardinality=0..*,summary"`
	// Sequence that was observed
	ObservedSeq *string `json:"observedSeq,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for ObservedSeq
	ObservedSeqExt *primitives.PrimitiveExtension `json:"_observedSeq,omitempty" fhir:"cardinality=0..1"`
	// An set of value as quality of sequence
	Quality []MolecularSequenceQuality `json:"quality,omitempty" fhir:"cardinality=0..*,summary"`
	// Average number of reads representing a given nucleotide in the reconstructed sequence
	ReadCoverage *int `json:"readCoverage,omitempty" fhir:"cardinality=0..1,summary"`
	// Extension for ReadCoverage
	ReadCoverageExt *primitives.PrimitiveExtension `json:"_readCoverage,omitempty" fhir:"cardinality=0..1"`
	// External repository which contains detailed report related with observedSeq in this resource
	Repository []MolecularSequenceRepository `json:"repository,omitempty" fhir:"cardinality=0..*,summary"`
	// Pointer to next atomic sequence
	Pointer []Reference `json:"pointer,omitempty" fhir:"cardinality=0..*,summary"`
	// Structural variant
	StructureVariant []MolecularSequenceStructureVariant `json:"structureVariant,omitempty" fhir:"cardinality=0..*,summary"`
}
